Cics SIGNON improvements

mm-broadcom opened the pull request:
## How Has This Been Tested?

Created and ran tests based upon the grammar and checks added.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Revised Grammar

mm-broadcom committed:
Grammar checks

mm-broadcom committed:
Add tests for SIGNON

mm-broadcom committed:
Added second invalid test

mm-broadcom commented:
@ilidio-lopes Could you review this?

mm-broadcom committed:
Merge branch 'development' into CICS-SIGNON-Improvements

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -727,10 +727,9 @@ cics_signal: SIGNAL (EVENT cics_data_value | FROMCHANNEL cics_data_value | FROM
-             LANGUAGECODE cics_data_value | NATLANG cics_data_value | LANGINUSE cics_data_area | NATLANGINUSE cics_data_area |
-             PASSWORD cics_data_value | NEWPASSWORD cics_data_value | PHRASE cics_data_area | PHRASELEN cics_data_value |
-             NEWPHRASE cics_data_area | NEWPHRASELEN cics_data_value | OIDCARD cics_data_value | cics_handle_response)*;
+cics_signon: SIGNON (cics_signon_body | cics_handle_response)+;
```
`cics_handle_response` is handled in the `cics_signon_body`
```suggestion
cics_signon: SIGNON cics_signon_body;
```

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSignonOptionsCheckUtility.java:
```diff
@@ -0,0 +1,95 @@
+    private void checkRule(CICSParser.Cics_signon_bodyContext ctx) {
+        checkHasMandatoryOptions(ctx.USERID(), ctx, "USERID");
+
+        if (!ctx.NEWPHRASE().isEmpty())
+            checkHasMandatoryOptions(ctx.NEWPHRASELEN(), ctx, "NEWPHRASELEN");
```
This and associated `checkPrerequisiteIsMet` could be handled by `checkAllOptionsArePresentOrAbsent` now.

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Added missing TOKEN subrule

- Added missing TOKEN subrule to the SIGNON grammar
- Added checks for TOKEN
- Added tests for TOKEN

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -727,10 +727,11 @@ cics_signal: SIGNAL (EVENT cics_data_value | FROMCHANNEL cics_data_value | FROM
+cics_signon: SIGNON (cics_signon_body|cics_signon_token_body);
+cics_signon_body: ((USERID | GROUPID | LANGUAGECODE | NATLANG | PASSWORD | NEWPASSWORD | PHRASELEN | NEWPHRASELEN | OIDCARD) cics_data_value |
+                    (CHANGETIME | DAYSLEFT | ESMREASON | ESMRESP | EXPIRYTIME | INVALIDCOUNT | LANGINUSE | LASTUSETIME | NATLANGINUSE | PHRASE | NEWPHRASE) cics_data_area | cics_handle_response)+;
+cics_signon_token_body: ((KERBEROS | KERBEROS | BIT | BIT | BASE64 | BASE64) | (TOKENLEN | GROUPID | GROUPID | LANGUAGECODE | NATLANG) cics_data_value |
+                    (TOKEN | LANGINUSE | LANGINUSE | NATLANGINUSE | NATLANGINUSE | ESMREASON | ESMREASON | ESMRESP | ESMRESP) cics_data_area | (TOKENTYPE | TOKENTYPE | DATATYPE) cics_cvda | cics_handle_response)+;
```
```suggestion
cics_signon_token_body: ((KERBEROS | BIT | BASE64) | (TOKENLEN | GROUPID | LANGUAGECODE | NATLANG) cics_data_value |
                    (TOKEN | LANGINUSE | NATLANGINUSE | ESMREASON | ESMRESP) cics_data_area | (TOKENTYPE | DATATYPE) cics_cvda | cics_handle_response)+;
```

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSignonOptionsCheckUtility.java:
```diff
@@ -0,0 +1,111 @@
+        checkMutuallyExclusiveOptions("BIT, DATATYPE or BASE64", ctx.BIT(), ctx.DATATYPE(), ctx.BASE64());
+        checkMutuallyExclusiveOptions("LANGUAGECODE or NATLANG", ctx.LANGUAGECODE(), ctx.NATLANG());
+
+        checkPrerequisiteIsMet(ctx.TOKEN(), ctx.TOKENLEN(), ctx, "TOKENLEN without TOKEN");
```
`TOKEN` and `TOKENLEN` are both mandatory.

mm-broadcom replied:
Fixed in latest commit.

mm-broadcom committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Add required values for SIGNON TOKEN

mm-broadcom committed:
Use variables in tests
